File: app/src/main/java/com/prance/teacher/services/UpdateService.java
package com.prance.teacher.services;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Build;
import android.os.IBinder;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v4.content.FileProvider;
import android.support.v4.content.LocalBroadcastManager;

import com.blankj.utilcode.util.LogUtils;
import com.blankj.utilcode.util.SDCardUtils;
import com.leo.download.DownloadError;
import com.leo.download.DownloadListener;
import com.leo.download.DownloadManager;

import java.io.File;

/**
 * Created by bingbing on 2016/10/20.
 */

public class UpdateService extends Service {

    public static final String DIR = "app_update";
    public static final String DOWNLOAD_BROADCAST_CANCEL_ACTION = "com.prance.lib.update.UpdateService.DownloadBroadcast.cancel";
    public static final String DOWNLOAD_BROADCAST_INSTALL_ACTION = "com.prance.lib.update.UpdateService.DownloadBroadcast.install";
    private DownloadBroadcast broadcast;
    private String url;
    private int notifyId = 20002;
    private File downloadFile = null;

    public static final String DOWNLOAD_BROADCAST_ACTIVITY_ACTION_DISMISS = "com.prance.lib.update.DownloadAlertActivity.dismiss";
    public static final String DOWNLOAD_BROADCAST_ACTIVITY_ACTION_PROGRESS = "com.prance.lib.update.DownloadAlertActivity.progress";
    public static final String DOWNLOAD_BROADCAST_ACTIVITY_ACTION_ERROR = "com.prance.lib.update.DownloadAlertActivity.error";

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        if (intent == null || intent.getExtras() == null) {
            return super.onStartCommand(intent, flags, startId);
        }
        String temp = intent.getExtras().getString("url");

        if (temp.equals(url)) {
            return super.onStartCommand(intent, flags, startId);
        } else {
            url = temp;
        }

        if (broadcast == null) {
            IntentFilter filter = new IntentFilter();
            filter.addAction(DOWNLOAD_BROADCAST_CANCEL_ACTION);
            filter.addAction(DOWNLOAD_BROADCAST_INSTALL_ACTION);
            registerReceiver(broadcast = new DownloadBroadcast(), filter);
        }

        File parentFile = new File(SDCardUtils.getSDCardPathByEnvironment()).getParentFile();
        File updateDir = new File(parentFile, DIR);

        if (!updateDir.exists()) {
            updateDir.mkdirs();
        }

        DownloadManager.getInstance(getApplicationContext()).enquene(url, updateDir.getPath(), new DownloadListener() {

            long currentTime = 0;

            @Override
            public void onStart(int id, long size) {
                LogUtils.d("onStart");
                currentTime = System.currentTimeMillis();

            }

            @Override
            public void onProgress(int id, long currSize, long totalSize) {
                if (System.currentTimeMillis() - currentTime >= 1000) {
                    currentTime = System.currentTimeMillis();


                }
            }

            @Override
            public void onRestart(int id, long currSize, long totalSize) {
                LogUtils.d("onRestart");
            }

            @Override
            public void onPause(int id, long currSize) {
                LogUtils.d("onPause");
            }

            @Override
            public void onComplete(int id, String dir, String name) {
                LogUtils.d("onComplete");

                downloadFile = new File(dir, name);
                url = null;


                NotificationManagerCompat
                        .from(getApplicationContext()).cancel(notifyId);

                LocalBroadcastManager.getInstance(UpdateService.this).sendBroadcast(new Intent(DOWNLOAD_BROADCAST_ACTIVITY_ACTION_DISMISS));

                Intent cancelIntent = new Intent(DOWNLOAD_BROADCAST_INSTALL_ACTION);
                sendBroadcast(cancelIntent);
            }

            @Override
            public void onCancel(int id) {
                LogUtils.d("onCancel");
            }

            @Override
            public void onError(int id, DownloadError error) {
                LogUtils.d("onError");
                url = null;

                Intent progressIntent = new Intent(DOWNLOAD_BROADCAST_ACTIVITY_ACTION_ERROR);
                LocalBroadcastManager.getInstance(UpdateService.this).sendBroadcast(progressIntent);
            }
        });

        return super.onStartCommand(intent, 0, startId);
    }

    class DownloadBroadcast extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            LogUtils.d(intent.getAction());

            try {
                if (intent.getAction().equals(DOWNLOAD_BROADCAST_CANCEL_ACTION)) {
                    DownloadManager.getInstance(context).cancel(url, null);
                    DownloadManager.getInstance(context).deleteRecord(url);
                    NotificationManagerCompat
                            .from(getApplicationContext()).cancel(notifyId);

                    LocalBroadcastManager.getInstance(UpdateService.this).sendBroadcast(new Intent(DOWNLOAD_BROADCAST_ACTIVITY_ACTION_DISMISS));

                    stopSelf();
                } else if (intent.getAction().equals(DOWNLOAD_BROADCAST_INSTALL_ACTION) && downloadFile != null) {
                    Intent install = new Intent(Intent.ACTION_VIEW);
                    install.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    File apkFile = downloadFile;
                    //判断是否是AndroidN以及更高的版本
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        install.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                        Uri contentUri = FileProvider.getUriForFile(context, context.getPackageName()+".fileprovider", apkFile);
                        install.setDataAndType(contentUri, "application/vnd.android.package-archive");
                    } else {
                        install.setDataAndType(Uri.fromFile(apkFile), "application/vnd.android.package-archive");
                    }
                    context.startActivity(install);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            url = null;
        }

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (broadcast != null) {
            unregisterReceiver(broadcast);
        }
        LocalBroadcastManager.getInstance(this).sendBroadcast(new Intent(DOWNLOAD_BROADCAST_ACTIVITY_ACTION_DISMISS));
    }
}

File: app/src/main/java/com/prance/teacher/services/UpdateService.kt
package com.prance.teacher.services

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Binder
import android.os.IBinder

import com.blankj.utilcode.util.LogUtils
import com.liulishuo.okdownload.DownloadTask
import com.prance.lib.common.utils.DiskPathUtils

import java.io.File
import com.liulishuo.okdownload.core.cause.ResumeFailedCause
import com.liulishuo.okdownload.core.listener.DownloadListener3
import com.prance.teacher.BuildConfig
import java.lang.Exception


/**
 * Created by bingbing on 2016/10/20.
 */

class UpdateService : Service() {

    private var url: String? = null
    private var downloadFile: File? = null
    private var mListener: MyDownloadListener? = null
    private var mTask: DownloadTask? = null
    //用于和外界交互
    private val binder = UpdateServiceBinder()

    inner class UpdateServiceBinder : Binder() {

        fun setListener(listener: MyDownloadListener) {
            this@UpdateService.mListener = listener
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {

        if (intent == null || intent.extras == null) {
            return super.onStartCommand(intent, flags, startId)
        }
        val temp = intent.extras!!.getString(URL)

        if (temp == url) {
            return super.onStartCommand(intent, flags, startId)
        } else {
            url = temp
        }

        val parentFile = File(DiskPathUtils.getDiskFileDir(this)).parentFile
        val updateDir = File(parentFile, DIR)

        if (!updateDir.exists()) {
            updateDir.mkdirs()
        }

        url?.run {

            if(!this.startsWith("http")){
                mListener?.onError()
                return@run
            }

            mTask = DownloadTask.Builder(this, updateDir)
                    .setFilename("""${this.hashCode()}.apk""")
                    // the minimal interval millisecond for callback progress
                    .setMinIntervalMillisCallbackProcess(50)
                    // do re-download even if the task has already been completed in the past.
                    .setPassIfAlreadyCompleted(true)
                    .build()

            mTask?.run {
                if (BuildConfig.DEBUG) {
//                    OkDownload.with().downloadDispatcher().cancel(id)
//                    OkDownload.with().breakpointStore().remove(id)
//                    file?.delete()
                }

                enqueue(object : DownloadListener3() {

                    override fun warn(task: DownloadTask) {

                    }

                    override fun started(task: DownloadTask) {
                        mListener?.onStartDownload()
                    }

                    override fun completed(task: DownloadTask) {
                        task.filename?.run {
                            mListener?.onComplete(task.parentFile.path, this)
                        }
                    }

                    override fun error(task: DownloadTask, e: Exception) {
                        mListener?.onError()
                    }

                    override fun canceled(task: DownloadTask) {

                    }

                    override fun connected(task: DownloadTask, blockCount: Int, currentOffset: Long, totalLength: Long) {

                    }

                    override fun retry(task: DownloadTask, cause: ResumeFailedCause) {

                    }

                    override fun progress(task: DownloadTask, currentOffset: Long, totalLength: Long) {
                        mListener?.onProgress(currentOffset, totalLength)
                    }
                })
            }

        }

        return super.onStartCommand(intent, 0, startId)
    }

    override fun onDestroy() {
        super.onDestroy()
        mTask?.cancel()
        LogUtils.d("onDestroy")
    }

    interface MyDownloadListener {
        fun onStartDownload() {}
        fun onProgress(currSize: Long, totalSize: Long)
        fun onComplete(dir: String, name: String)
        fun onError()
    }

    companion object {

        const val DIR = "app_update"
        const val URL = "url"

        fun callingIntent(context: Context, url: String): Intent {
            val intent = Intent(context, UpdateService::class.java)
            intent.putExtra(URL, url)
            return intent
        }

        fun callingIntent(context: Context): Intent {
            return Intent(context, UpdateService::class.java)
        }
    }
}
